docs/design/federated-services.md

```diff
@@ -0,0 +1,490 @@
+# Kubernetes Cluster Federation (a.k.a. "Ubernetes")
+## Cross-cluster Load Balancing and Service Discovery 
+### Requirements and System Design
+### by Quinton Hoole, Dec 3 2015
+
+## Requirements
+
+### Discovery, Load-balancing and Failover
+
+1. **Internal discovery and connection**: Pods/containers (running in
+   a Kubernetes cluster) must be able to easily discover and connect
+   to endpoints for Kubernetes services on which they depend in a
+   consistent way, irrespective of whether those services exist in a
+   different kubernetes cluster within the same cluster federation.
+   Hence-forth referred to as "cluster-internal clients", or simply
+   "internal clients".
+1. **External discovery and connection**: External clients (running
+   outside a Kubernetes cluster) must be able to discover and connect
+   to endpoints for Kubernetes services on which they depend.
+   1. **External clients predominantly speak HTTP(S)**: External
+      clients are most often, but not always, web browsers, or at
+      least speak HTTP(S) (TBD list notable exceptions.  DNS servers?
+      SIP servers?  Database Servers?)
+1. **Find the "best" endpoint:** Upon initial discovery and
```

Comment:
Yes, RTT is a good proxy for distance in many cases, and so the implementation might be simply that.  But there are actually multiple reasons why "close" is good, beyond pure average latency reasons like the TLS example you cite.  For example, bandwidth charges ($$) within a zone are typically zero, whereas across zones they can be very substantial (etune@ once did a calc to demonstrate how they can easily outweigh CPU charges by several orders of magnitude. Also, it's well-known fact that multi-tier distributed applications suffer disproportionately at the tail end of latency (e.g. a frontend often needs to wait for the slowest of all of it's backends, transitively, so the 99th percentile latency of the backends can easily become the median of the frontend).  Here's a pretty good article describing the phenomenon http://highscalability.com/blog/2012/3/12/google-taming-the-long-latency-tail-when-more-machines-equal.html.